File: Task/Latin-Squares-in-reduced-form-Randomizing-using-Jacobson-and-Matthews-technique/C-sharp/latin-squares-in-reduced-form-randomizing-using-jacobson-and-matthews-technique.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

public sealed class LatinSquaresInReducedForm
{
    private static readonly Random random = new Random();

    public static void Main(string[] args)
    {
        Console.WriteLine("PART 1: 10,000 latin Squares of order 4 in reduced form:\n");
        int[,] original4 = new int[,] { { 1, 2, 3, 4 }, { 2, 1, 4, 3 }, { 3, 4, 1, 2 }, { 4, 3, 2, 1 } };
        Dictionary<string, int> frequencies = new Dictionary<string, int>();
        int[,,] cube = CreateCube(original4, 4);

        for (int i = 1; i <= 10_000; i++)
        {
            ShuffleCube(cube);
            int[,] matrix = ToMatrix(cube);
            Reduce(matrix);
            OneBased(matrix);
            string key = MatrixToString(matrix);
            if (frequencies.ContainsKey(key))
                frequencies[key]++;
            else
                frequencies[key] = 1;
        }

        foreach (var entry in frequencies)
        {
            Console.Write(entry.Key);
            Console.WriteLine($" occurs {entry.Value} times");
        }

        Console.WriteLine("\nPART 2: 10_000 latin squares of order 5 in reduced form:");
        int[,] original5 = new int[,] {
            { 1, 2, 3, 4, 5 },
            { 2, 3, 4, 5, 1 },
            { 3, 4, 5, 1, 2 },
            { 4, 5, 1, 2, 3 },
            { 5, 1, 2, 3, 4 }
        };
        frequencies.Clear();
        cube = CreateCube(original5, 5);

        for (int i = 1; i <= 10_000; i++)
        {
            ShuffleCube(cube);
            int[,] matrix = ToMatrix(cube);
            Reduce(matrix);
            string key = MatrixToString(matrix);
            if (frequencies.ContainsKey(key))
                frequencies[key]++;
            else
                frequencies[key] = 1;
        }

        int count = 0;
        foreach (int frequency in frequencies.Values)
        {
            count++;
            Console.Write($"{(count > 1 ? ", " : "")}{(count % 8 == 1 ? "\n" : "")}{count,2}({frequency,3})");
        }

        Console.WriteLine("\n\nPART 3: 750 latin squares of order 42, showing the last one:\n");
        int[,] matrix42 = null;
        cube = CreateCube(null, 42);
        for (int i = 1; i <= 750; i++)
        {
            ShuffleCube(cube);
            if (i == 750)
            {
                matrix42 = ToMatrix(cube);
                OneBased(matrix42);
            }
        }
        PrintMatrix(matrix42);

        Console.WriteLine("\nPART 4: 1,000 latin squares of order 256:\n");
        Stopwatch stopwatch = Stopwatch.StartNew();
        cube = CreateCube(null, 256);
        for (int i = 1; i <= 1_000; i++)
        {
            ShuffleCube(cube);
        }
        stopwatch.Stop();
        Console.WriteLine($"Generated in {stopwatch.ElapsedMilliseconds} milliseconds");
    }

    private static void Reduce(int[,] matrix)
    {
        int size = matrix.GetLength(0);

        for (int j = 0; j < size - 1; j++)
        {
            if (matrix[0, j] != j)
            {
                for (int k = j + 1; k < size; k++)
                {
                    if (matrix[0, k] == j)
                    {
                        for (int i = 0; i < size; i++)
                        {
                            int temp = matrix[i, j];
                            matrix[i, j] = matrix[i, k];
                            matrix[i, k] = temp;
                        }
                        break;
                    }
                }
            }
        }

        for (int i = 1; i < size - 1; i++)
        {
            if (matrix[i, 0] != i)
            {
                for (int k = i + 1; k < size; k++)
                {
                    if (matrix[k, 0] == i)
                    {
                        for (int j = 0; j < size; j++)
                        {
                            int temp = matrix[i, j];
                            matrix[i, j] = matrix[k, j];
                            matrix[k, j] = temp;
                        }
                        break;
                    }
                }
            }
        }
    }

    private static int[,] ToMatrix(int[,,] cube)
    {
        int size = cube.GetLength(0);
        int[,] matrix = new int[size, size];
        for (int i = 0; i < size; i++)
        {
            for (int j = 0; j < size; j++)
            {
                for (int k = 0; k < size; k++)
                {
                    if (cube[i, j, k] != 0)
                    {
                        matrix[i, j] = k;
                        break;
                    }
                }
            }
        }
        return matrix;
    }

    private static void ShuffleCube(int[,,] cube)
    {
        int size = cube.GetLength(0);
        bool proper = true;

        int rx, ry, rz;
        do
        {
            rx = random.Next(0, size);
            ry = random.Next(0, size);
            rz = random.Next(0, size);
        } while (cube[rx, ry, rz] != 0);

        while (true)
        {
            int ox = 0, oy = 0, oz = 0;

            while (cube[ox, ry, rz] != 1)
            {
                ox++;
            }
            while (cube[rx, oy, rz] != 1)
            {
                oy++;
            }
            while (cube[rx, ry, oz] != 1)
            {
                oz++;
            }

            if (!proper)
            {
                if (random.Next(2) == 0)
                {
                    ox++;
                    while (cube[ox, ry, rz] != 1)
                    {
                        ox++;
                    }
                }
                if (random.Next(2) == 0)
                {
                    oy++;
                    while (cube[rx, oy, rz] != 1)
                    {
                        oy++;
                    }
                }
                if (random.Next(2) == 0)
                {
                    oz++;
                    while (cube[rx, ry, oz] != 1)
                    {
                        oz++;
                    }
                }
            }

            cube[rx, ry, rz]++;
            cube[rx, oy, oz]++;
            cube[ox, ry, oz]++;
            cube[ox, oy, rz]++;

            cube[rx, ry, oz]--;
            cube[rx, oy, rz]--;
            cube[ox, ry, rz]--;
            cube[ox, oy, oz]--;

            if (cube[ox, oy, oz] < 0)
            {
                rx = ox; ry = oy; rz = oz;
                proper = false;
            }
            else
            {
                break;
            }
        }
    }

    private static int[,,] CreateCube(int[,] matrix, int size)
    {
        int[,,] cube = new int[size, size, size];
        for (int i = 0; i < size; i++)
        {
            for (int j = 0; j < size; j++)
            {
                int k = (matrix == null) ? (i + j) % size : matrix[i, j] - 1;
                cube[i, j, k] = 1;
            }
        }
        return cube;
    }

    private static void OneBased(int[,] matrix)
    {
        int rows = matrix.GetLength(0);
        int cols = matrix.GetLength(1);
        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                matrix[i, j]++;
            }
        }
    }

    private static string MatrixToString(int[,] matrix)
    {
        int rows = matrix.GetLength(0);
        int cols = matrix.GetLength(1);
        var result = new System.Text.StringBuilder();
        result.Append("[");
        for (int i = 0; i < rows; i++)
        {
            result.Append("[");
            for (int j = 0; j < cols; j++)
            {
                result.Append(matrix[i, j]);
                if (j < cols - 1) result.Append(", ");
            }
            result.Append("]");
            if (i < rows - 1) result.Append(", ");
        }
        result.Append("]");
        return result.ToString();
    }

    private static void PrintMatrix(int[,] matrix)
    {
        int rows = matrix.GetLength(0);
        int cols = matrix.GetLength(1);
        for (int i = 0; i < rows; i++)
        {
            Console.Write("[");
            for (int j = 0; j < cols; j++)
            {
                Console.Write(matrix[i, j]);
                if (j < cols - 1) Console.Write(", ");
            }
            Console.WriteLine("]");
        }
    }
}

File: Task/Latin-Squares-in-reduced-form-Randomizing-using-Jacobson-and-Matthews-technique/C++/latin-squares-in-reduced-form-randomizing-using-jacobson-and-matthews-technique.cpp
#include <algorithm>
#include <chrono>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <map>
#include <random>
#include <vector>

std::random_device random;
std::mt19937 generator(random());

template <typename T>
void print_one_based_vector(const std::vector<T>& list) {
	std::cout << "[";
	for ( uint32_t i = 0; i < list.size() - 1; ++i ) {
		std::cout << list[i] + 1 << ", ";
	}
	std::cout << list.back() + 1 << "]";
}

template <typename T>
void print_2D_vector(const std::vector<std::vector<T>>& lists) {
	std::cout << "[";
	for ( uint32_t i = 0; i < lists.size() - 1; ++i ) {
		print_one_based_vector(lists[i]); std::cout << ", ";
	}
	print_one_based_vector(lists.back()); std::cout << "]";
}

std::vector<std::vector<std::vector<int32_t>>> create_cube(const std::vector<std::vector<int32_t>>& matrix,
														   const uint32_t& size) {
	std::vector<std::vector<std::vector<int32_t>>> cube =
		{ size, std::vector<std::vector<int32_t>>(size, std::vector<int32_t>(size, 0)) };
	for ( uint32_t i = 0; i < size; ++i ) {
		for ( uint32_t j = 0; j < size; ++j ) {
			const int32_t k = ( matrix.empty() ) ? ( i + j ) % size : matrix[i][j] - 1;
			cube[i][j][k] = 1;
		}
	}
	return cube;
}

void shuffle_cube(std::vector<std::vector<std::vector<int32_t>>>& cube) {
	bool proper = true;

	std::uniform_int_distribution<int32_t> distribution(0, cube.size() - 1);
	uint32_t rx, ry, rz;
	do {
		rx = distribution(generator);
		ry = distribution(generator);
		rz = distribution(generator);
	} while ( cube[rx][ry][rz] != 0 );

	while ( true ) {
		uint32_t ox = 0, oy = 0, oz = 0;

		while ( cube[ox][ry][rz] != 1 ) {
			ox++;
		}
		while ( cube[rx][oy][rz] != 1 ) {
			oy++;
		}
		while ( cube[rx][ry][oz] != 1 ) {
			oz++;
		}

		std::uniform_int_distribution<int32_t> distribution(0, 1);
		if ( ! proper ) {
			if ( distribution(generator) == 0 ) {
				ox++;
				while ( cube[ox][ry][rz] != 1 ) {
					ox++;
				}
			}
			if ( distribution(generator) == 0 ) {
				oy++;
				while ( cube[rx][oy][rz] != 1 ) {
					oy++;
				}
			}
			if ( distribution(generator) == 0 ) {
				oz++;
				while ( cube[rx][ry][oz] != 1 ) {
					oz++;
				}
			}
		}

		cube[rx][ry][rz]++;
		cube[rx][oy][oz]++;
		cube[ox][ry][oz]++;
		cube[ox][oy][rz]++;

		cube[rx][ry][oz]--;
		cube[rx][oy][rz]--;
		cube[ox][ry][rz]--;
		cube[ox][oy][oz]--;

		if ( cube[ox][oy][oz] < 0 ) {
			rx = ox; ry = oy; rz = oz;
			proper = false;
		} else {
			break;
		}
	}
}

std::vector<std::vector<int32_t>> to_matrix(const std::vector<std::vector<std::vector<int32_t>>>& cube) {
	std::vector<std::vector<int32_t>> matrix = { cube.size(), std::vector<int32_t>(cube.size(), 0) };
	for ( uint32_t i = 0; i < cube.size(); ++i ) {
		for ( uint32_t j = 0; j < cube.size(); ++j ) {
			for ( uint32_t k = 0; k < cube.size(); ++k ) {
				if ( cube[i][j][k] != 0 ) {
					matrix[i][j] = k;
					break;
				}
			}
		}
	}
	return matrix;
}

void reduce(std::vector<std::vector<int32_t>>& matrix) {
	for ( uint32_t j = 0; j < matrix.size() - 1; ++j ) {
		if ( matrix[0][j] != static_cast<int32_t>(j) ) {
			for ( uint32_t k = j + 1; k < matrix.size(); ++k ) {
				if ( matrix[0][k] == static_cast<int32_t>(j) ) {
					for ( uint32_t i = 0; i < matrix.size(); ++i ) {
						std::swap(matrix[i][j], matrix[i][k]);
					}
					break;
				}
			}
		}
	}

	for ( uint32_t i = 1; i < matrix.size() - 1; ++i ) {
		if ( matrix[i][0] != static_cast<int32_t>(i) ) {
			for ( uint32_t k = i + 1; k < matrix.size(); ++k ) {
				if ( matrix[k][0] == static_cast<int32_t>(i) ) {
					for ( uint32_t j = 0; j < matrix.size(); ++j ) {
						std::swap(matrix[i][j], matrix[k][j]);
					}
					break;
				}
			}
		}
	}
}

int main() {
	std::cout << "PART 1: 10,000 latin Squares of order 4 in reduced form:" << std::endl << std::endl;
	std::vector<std::vector<int32_t>> original_4 =
		{ { 1, 2, 3, 4 }, { 2, 1, 4, 3 }, { 3, 4, 1, 2 }, { 4, 3, 2, 1 } };
	std::map<std::vector<std::vector<int32_t>>, uint32_t> frequencies{ };
	std::vector<std::vector<std::vector<int32_t>>> cube = create_cube(original_4, 4);
	for ( uint32_t i = 1; i <= 10'000; ++i ) {
		shuffle_cube(cube);
		std::vector<std::vector<int32_t>> matrix = to_matrix(cube);
		reduce(matrix);
		frequencies[matrix]++;
	}

	for ( std::pair<std::vector<std::vector<int32_t>>, uint32_t> pair : frequencies ) {
		print_2D_vector(pair.first);
		std::cout << " occurs " << pair.second << " times" << std::endl;
	}

	std::cout << "\n" << "PART 2: 10_000 latin squares of order 5 in reduced form:" << std::endl;
		std::vector<std::vector<int32_t>> original_5 = { { 1, 2, 3, 4, 5 }, { 2, 3, 4, 5, 1 },
			                          { 3, 4, 5, 1, 2 }, { 4, 5, 1, 2, 3 }, { 5, 1, 2, 3, 4 } };
		frequencies.clear();
		cube = create_cube(original_5, 5);
		for ( uint32_t i = 1; i <= 10'000; ++i ) {
			shuffle_cube(cube);
			std::vector<std::vector<int32_t>> matrix = to_matrix(cube);
			reduce(matrix);
			frequencies[matrix]++;
		}

		uint32_t count = 0;
		for ( std::pair<std::vector<std::vector<int32_t>>, uint32_t> pair : frequencies ) {
			count++;
			std::cout << ( count > 1 ? ", " : "" ) << ( count % 8 == 1 ? "\n" : "" ) << std::setw(2) << count
					  << "(" << std::setw(3) << pair.second << ")";
		}

		std::cout << "\n\n"
				  << "PART 3: 750 latin squares of order 42, showing the last one:" << "\n" << std::endl;
		std::vector<std::vector<int32_t>> matrix_42{ };
		cube = create_cube(matrix_42, 42);
		for ( uint32_t i = 1; i <= 750; ++i ) {
			shuffle_cube(cube);
			if ( i == 750 ) {
				matrix_42 = to_matrix(cube);
			}
		}

		for ( const std::vector<int32_t>& row : matrix_42 ) {
			print_one_based_vector(row);
			std::cout << std::endl;
		}

		std::cout << "\n" << "PART 4: 1,000 latin squares of order 256:" << "\n" << std::endl;
		const auto start = std::chrono::steady_clock::now();
		std::vector<std::vector<int32_t>> empty{ };
		cube = create_cube(empty, 256);
		for ( uint32_t i = 1; i <= 1'000; ++i ) {
			shuffle_cube(cube);
		}
		const auto finish = std::chrono::steady_clock::now();
		std::cout << "Generated in " << std::chrono::duration<double, std::milli>( finish - start ).count()
				  << " milliseconds" << std::endl;
}
